test/spec/features/service/exec_spec.rb

```diff
@@ -36,22 +36,77 @@
     expect(k.code).to eq(126)
   end
 
-  it 'runs a command inside a service with tty' do
-    k = kommando("kontena service exec -it test-1 sh")
+  describe '--tty' do
+    it 'runs a command inside a service' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        k.in << "ls -la /\r"
+        k.out.on "lib64" do
+          k.in << "exit\r"
+        end
+      end
+      expect(k.run).to be_truthy
+    end
+
+    it 'runs a command with tty control input' do
+      k = kommando("kontena service exec -it test-1 sh")
 
-    k.out.on("#") do
-      k.in << "ls -la /\r"
-      k.out.on "lib64" do
-        k.in << "exit\r"
+      k.out.on("#") do
+        k.in << "sleep 10 && echo ok\r"
+        sleep 0.1
+        k.in << "\x03" # Ctrl-C -> SIGINT
+        sleep 0.1
+        k.in << "\x04" # Ctrl-D -> EOF
       end
+
+      expect(k.run).to be_truthy
+      expect(k.out).to match /\^C/
+    end
+
+    it 'runs a command with non-ascii input' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        k.in << "echo f\u00e5\u00e5 | LANG=C.UTF-8 rev\r"
+        k.out.on("#") do
+          k.in << "exit\r"
+        end
+      end
+
+      expect(k.run).to be_truthy
+      expect(k.out).to match /\u00e5\u00e5f/
+    end
+
+    # RuntimeError : stdin read JSON::GeneratorError: partial character in source, but hit end
+    pending 'runs a command with UTF-8 input across block boundries' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        # with an odd number of bytes in the 'echo ', the 2-byte UTF-8 chars will get split across the 1024-byte chunk boundary
```

Comment:
> This will fail if the input is not actually valid UTF-8, or if the io.readpartial(1024) chunk boundary splits a valid UTF-8 character.

Yes, easy enough to repro like this:

```

service exec
 [done] Creating test-1 service      
 [done] Deploying service test-1      
⊛ Deployed instance e2e/null/test-1-1 to node localhost
  --tty
#  [error] RuntimeError : stdin read JSON::GeneratorError: partial character in source, but hit end
         See /root/.kontena/kontena.log or run the command again with environment DEBUG=true set to see the full exception
    runs a command with UTF-8 input across block boundries (PENDING: No reason given)
 [done] Removing service test-1      

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) service exec --tty runs a command with UTF-8 input across block boundries
     # No reason given
     Failure/Error: expect(k.out).to match /^\u00e5{512}$/
     
       expected "#  [\e[31merror\e[0m] RuntimeError : stdin read JSON::GeneratorError: partial character in source, b.../kontena.log or run the command again with environment DEBUG=true set to see the full exception\r\n" to match /^\u00e5{512}$/
       Diff:
       @@ -1,2 +1,3 @@
       -/^\u00e5{512}$/
       +#  [error] RuntimeError : stdin read JSON::GeneratorError: partial character in source, but hit end
       +         See /root/.kontena/kontena.log or run the command again with environment DEBUG=true set to see the full exception
       
     # ./spec/features/service/exec_spec.rb:93:in `block (3 levels) in <top (required)>'
```